src/lib/pubkey/pqcrystals/pqcrystals_encoding.h

```diff
@@ -0,0 +1,228 @@
+ * forcing the compiler to generate specialized code for each supported range.
+ */
+template <int32_t range, crystals_trait PolyTrait>
+struct BitPackingTrait final {
+      using T = typename PolyTrait::T;
+      using unsigned_T = std::make_unsigned_t<T>;
+      using sink_t = uint64_t;
+
+      static_assert(range <= std::numeric_limits<T>::max());
+
+      constexpr static size_t bits_in_collector = sizeof(sink_t) * 8;
+      constexpr static size_t bits_per_coeff = bitlen(range);
+      constexpr static size_t bits_per_pack = [] {
+         // Ensure that the bit-packing is byte-aligned and scale it
+         // to utilize the collector's bit-width as much as possible.
+         size_t smallest_aligned_pack = std::lcm(bits_per_coeff, size_t(8));
+         return (smallest_aligned_pack < bits_in_collector)
+                   ? (bits_in_collector / smallest_aligned_pack) * smallest_aligned_pack
+                   : smallest_aligned_pack;
+      }();
+      constexpr static size_t bytes_per_pack = bits_per_pack / 8;
+      constexpr static size_t coeffs_per_pack = bits_per_pack / bits_per_coeff;
+      constexpr static size_t collectors_per_pack = (bytes_per_pack + sizeof(sink_t) - 1) / sizeof(sink_t);
+      constexpr static size_t collector_bytes_per_pack = collectors_per_pack * sizeof(sink_t);
+      constexpr static sink_t value_mask = (1 << bits_per_coeff) - 1;
+
+      using collector_array = std::array<sink_t, collectors_per_pack>;
+      using collector_bytearray = std::array<uint8_t, collector_bytes_per_pack>;
+
+      static_assert(PolyTrait::N % coeffs_per_pack == 0);
+};
+
+/**
+ * Base implementation of NIST FIPS 203 IPD Algorithm 4 (ByteEncode) and NIST
+ * FIPS 204 Algorithms 10 (SimpleBitPack) and 11 (BitPack).
+ *
+ * This takes a polynomial @p p and packs its coefficients into the buffer
+ * represented by @p stuffer. Optionally, the coefficients can be transformed
+ * using the @p map function before packing them. Kyber uses @p map to compress
+ * the coefficients as needed, Dilithium to transform coefficients to unsigned.
+ *
+ * The implementation assumes that the values returned from the custom @p map
+ * transformation are in the range [0, range]. No assumption is made about the
+ * value range of the coefficients in the polynomial @p p.
+ *
+ * Note that this bit-packing algorithm is inefficient if the bit-length of the
+ * coefficients is a multiple of 8. In that case, a byte-level encoding (that
+ * might need to take endianess into account) would be more efficient. However,
+ * neither Kyber nor Dilithium instantiate bit-packings with such a value range.
+ *
+ * @tparam range the upper bound of the coefficient range.
+ */
+template <int32_t range, crystals_trait PolyTrait, Domain D, coeff_map_fn<typename PolyTrait::T> MapFnT>
+constexpr void pack(const Polynomial<PolyTrait, D>& p, BufferStuffer& stuffer, MapFnT map) {
+   using trait = BitPackingTrait<range, PolyTrait>;
```

Comment:
Could the gist of this be useful beyond the CRYSTALS scope? E.g. FrodoKEM has a pack/unpack routine that -- on first glance at least -- looks like it could use it as well. @randombit perhaps there are more examples?

https://github.com/randombit/botan/blob/3657a72dfe697b4e56f0f0192bbf62cc164be2d6/src/lib/pubkey/frodokem/frodokem_common/frodo_matrix.cpp#L340-L384